### ISSUE TITLE
Support pluggable adapters with associated URL schemes and support pluggable XMLHTTPRequest handlers as an option argument for AJAX

### ISSUE THREAD

**Issue by yaronyg** [pouchdb/pouchdb#1453]
**tldr - We need to be able to submit our own adapter which would be called when a PouchDB object is created with our URL scheme. We also need a way to submit our own XMLHTTPRequest object to deps/ajax.js.**
## Background

I am working on a project called [Thali](https://thali.codeplex.com) that is looking at what it would take to build a peer to peer web. The core of the project is the CouchDB protocol. But because we are peer to peer we do not use HTTPS since it is based on a centralized security infrastructure (CAs). Instead we have our own scheme that we called [httpkey](http://thali.cloudapp.net/mediawiki/index.php?title=Httpkey_URL_Scheme). The core of httpkey is that identities in Thali _are_ public keys. So when two Thali peers talk they establish a TLS connection and use TLS mutual auth to exchange self signed certs. During cert validation they just want to validate that the key used to do the handshake is the key advertised in the X.509 cert.
## Relevance to PouchDB

Thali is intended to run across phones, tablets, desktops, etc. So we tend to use HTML/JS/CSS as our UX layer since we can use it everywhere. We are planning on using PouchDB as our client to talk to our CouchDB server (which is built using [CouchBase Lite](https://github.com/couchbase/couchbase-lite-java-core)) which runs on the local device.
## Today's hack

However Javascript can't (currently) support the crypto and TLS config functionality we need. So the way we make this work today is we take over XMLHTTPRequest and have PouchDB talk to [our hacked up XMLHTTPRequest](http://thali.codeplex.com/SourceControl/latest#Production/Utilities/DotNetUtilities/DotNetUtilities/ChromeNativeMessagingHost/ChromeExtension/xmlhttprequesttothali.js) which currently uses [our Chrome Extension](http://thali.codeplex.com/SourceControl/latest#Production/Utilities/DotNetUtilities/DotNetUtilities/ChromeNativeMessagingHost/ChromeExtension/) to talk to [our Chrome Native Host](http://thali.codeplex.com/SourceControl/latest#Production/Utilities/DotNetUtilities/DotNetUtilities/ChromeNativeMessagingHost/) which then handles all the [crypto magic](http://thali.codeplex.com/SourceControl/latest#Production/Utilities/DotNetUtilities/DotNetUtilities/DotNetUtilities/ThaliClientToDeviceHubUtilities.cs) needed to make this work.

We will also hook up the same XMLHTTPRequest hack to our [generic WebView bridge](http://thali.codeplex.com/SourceControl/latest#Production/Utilities/UniversalUtilities/src/main/java/com/msopentech/thali/utilities/webviewbridge/) that we use in both Android and JavaFX.

However we still can't pass through non-http URLs. So what we do is take our httpkey URL and turn it into a https URL. Then on the back end we look at the URLs before we process them to try and test to see if they are really a httpkey URL in disguise. If so then we transform it into a httpkey URL and continue processing from there.
### Pluggable adapters

If I'm reading the PouchDB code right then it seems like the obvious way to fix all of this is to provide a mechanism where by one can register both an adapter and a 'name' matching function (which I think would be called in setup/parseAdapter) that would check to see if a database name should be handled by that adapter.
### Adding an option to deps/ajax to handle custom XMLHTTPRequest objects

As mentioned above we really don't like taking over XMLHTTPRequest for everyone. Now, it's true that we could play games like simultaneously manage both our own XMLHTTPRequest object as well as the browser's until OPEN is called and we know which one to actually use. But this is really error prone. A much cleaner solution would be if we had an option for deps/ajax.js where we could submit our own XMLHTTPRequest object. If the option is set then rather than ajax just grabbing XMLHttpRequest it would check to see if the option is set and if so use the submitted XMLHTTPRequest object otherwise it would use the default.
## Conclusion

Both of these changes are actually pretty minor to the PouchDB code base. Literally a few lines. And they wouldn't break any existing code. I'm happy to submit a request with this proposal if there is interest but I first wanted to see if the group had any interest in taking these changes.

Thanks!

Yaron
